feat(ci): Add support for managing Halyard's versions.yml

kskewes-sf opened the pull request:
versions.yml:
- fetch versions.yml from Halyard's GCS bucket
- update versions.yml with new Spinnaker Version
  - replace any previous patch, eg: `1.27.1` will replace `1.27.0`
  - add any new minor, eg: `1.28.0`
  - keep at most 3 minor variations, eg: 1.27.1, 1.26.7, 1.25.6
  - validate versions.yml conforms to schema (Halyard java)
- publish versions.yml to Halyard's GCS bucket

additional:
- enable all Test classes to patch methods. This means we can use a base Test class instead of having to use a Git specific test class that does setup/teardown tasks that aren't needed.
- show in docs how to bypass local git config which can interfere with test runs (git commit/tag editor interaction)

kskewes-sf committed:
feat(ci): Add versions.yml commands

- fetch versions.yml from Halyard's GCS bucket
- update versions.yml with new Spinnaker Version
  - replace any previous patch, eg: 1.27.1 will replace 1.27.0
  - add any new minor, eg: 1.28.0
  - keep at most 3 minor variations, eg: 1.27.1, 1.26.7, 1.25.6
  - validate versions.yml conforms to schema
- publish versions.yml to Halyard's GCS bucket

kskewes-sf committed:
fix(ci): Enable patching methods in all tests - not just git

kskewes-sf committed:
fix(ci): Avoid personal git config interfering with tests

kskewes-sf commented:
```
$ ./dev/buildtool.sh fetch_versions
W 08:47:44.324 [MainThread.2229349] Monitoring is disabled
I 08:47:44.325 [MainThread.2229349] Fetching Spinnaker versions from gs://halconfig/versions.yml
I 08:47:50.345 [MainThread.2229349] Wrote Spinnaker versions to output/fetch_versions/versions.yml
W 08:47:50.345 [MainThread.2229349] Monitoring disabled -- dont push incremental metrics.
W 08:47:50.345 [MainThread.2229349] Monitoring disabled -- dont push final metrics.
```

```
$ cat output/fetch_versions/versions.yml 
illegalVersions:
- reason: Broken apache config makes the UI unreachable
  version: 1.2.0
- reason: UI does not load
  version: 1.4.0
latestHalyard: 1.51.0
latestSpinnaker: 1.29.0
versions:
- alias: v1.29.0
  changelog: https://spinnaker.io/changelogs/1.29.0-changelog/
  lastUpdate: 1664916719700
  minimumHalyardVersion: '1.45'
  version: 1.29.0
- alias: v1.28.1
  changelog: https://spinnaker.io/changelogs/1.28.1-changelog/
  lastUpdate: 1661279166000
  minimumHalyardVersion: '1.45'
  version: 1.28.1
- alias: v1.27.1
  changelog: https://spinnaker.io/changelogs/1.27.1-changelog/
  lastUpdate: 1660858660000
  minimumHalyardVersion: '1.45'
  version: 1.27.1
- alias: v1.26.7
  changelog: https://gist.github.com/spinnaker-release/e3714a97bbdd3e7c3b4d92adec938e7f
  lastUpdate: 1625100544812
  minimumHalyardVersion: '1.41'
  version: 1.26.7
```

kskewes-sf commented:
Add new patch version:
```
$ ./dev/buildtool.sh update_versions \ 
  --spinnaker_version 1.28.2 \
  --minimum_halyard_version '1.46' \
  --versions_yml_path output/fetch_versions/versions.yml 
W 08:48:40.843 [MainThread.2229693] Monitoring is disabled
I 08:48:40.847 [MainThread.2229693] Replacing version 1.28.1 with new version 1.28.2
I 08:48:40.849 [MainThread.2229693] Wrote Spinnaker versions to output/update_versions/versions.yml
W 08:48:40.849 [MainThread.2229693] Monitoring disabled -- dont push incremental metrics.
W 08:48:40.849 [MainThread.2229693] Monitoring disabled -- dont push final metrics.
```
```
$ cat output/update_versions/versions.yml 
illegalVersions:
- reason: Broken apache config makes the UI unreachable
  version: 1.2.0
- reason: UI does not load
  version: 1.4.0
latestHalyard: 1.51.0
latestSpinnaker: 1.29.0
versions:
- alias: v1.29.0
  changelog: https://spinnaker.io/changelogs/1.29.0-changelog/
  lastUpdate: 1664916719700
  minimumHalyardVersion: '1.45'
  version: 1.29.0
- alias: v1.28.2
  changelog: https://spinnaker.io/changelogs/1.28.2-changelog/
  lastUpdate: 1664999320847
  minimumHalyardVersion: '1.46'
  version: 1.28.2
- alias: v1.27.1
  changelog: https://spinnaker.io/changelogs/1.27.1-changelog/
  lastUpdate: 1660858660000
  minimumHalyardVersion: '1.45'
  version: 1.27.1
```

kskewes-sf commented:
Add new minor version
```
$ ./dev/buildtool.sh update_versions \
  --spinnaker_version 1.30.0 \
  --minimum_halyard_version '1.46' \
  --versions_yml_path output/fetch_versions/versions.yml 
W 08:49:43.432 [MainThread.2230062] Monitoring is disabled
I 08:49:43.437 [MainThread.2230062] Wrote Spinnaker versions to output/update_versions/versions.yml
W 08:49:43.437 [MainThread.2230062] Monitoring disabled -- dont push incremental metrics.
W 08:49:43.437 [MainThread.2230062] Monitoring disabled -- dont push final metrics.
```

```
$ cat output/update_versions/versions.yml illegalVersions:
- reason: Broken apache config makes the UI unreachable
  version: 1.2.0
- reason: UI does not load
  version: 1.4.0
latestHalyard: 1.51.0
latestSpinnaker: 1.30.0
versions:
- alias: v1.30.0
  changelog: https://spinnaker.io/changelogs/1.30.0-changelog/
  lastUpdate: 1664999383435
  minimumHalyardVersion: '1.46'
  version: 1.30.0
- alias: v1.29.0
  changelog: https://spinnaker.io/changelogs/1.29.0-changelog/
  lastUpdate: 1664916719700
  minimumHalyardVersion: '1.45'
  version: 1.29.0
- alias: v1.28.1
  changelog: https://spinnaker.io/changelogs/1.28.1-changelog/
  lastUpdate: 1661279166000
  minimumHalyardVersion: '1.45'
  version: 1.28.1
```

dbyron-sf commented on README.md:
```diff
@@ -394,3 +396,80 @@ The previous step created a branch in a local copy of your fork here:
+### Fetch versions.yml
+
+Halyard uses `versions.yml` as the source of truth for available Spinnaker
+versions to deploy. Fetch the file from GCS bucket for editing.
```
```suggestion
versions to deploy. Fetch the file from GCS for editing.
```
or maybe "its GCS bucket" ?

dbyron-sf commented on README.md:
```diff
@@ -394,3 +396,80 @@ The previous step created a branch in a local copy of your fork here:
+
+### Update versions.yml
+
+Add the new Spinnaker release version to `versions.yml`.
```
```suggestion
Add a new Spinnaker version to `versions.yml`.
```

dbyron-sf commented on README.md:
```diff
@@ -394,3 +396,80 @@ The previous step created a branch in a local copy of your fork here:
+
+### Publish versions.yml
+
+When ready to release the new Spinnaker version to users, publish the
```
```suggestion
When ready to release a new Spinnaker version to users, publish the
```

dbyron-sf commented on README.md:
```diff
@@ -394,3 +396,80 @@ The previous step created a branch in a local copy of your fork here:
+### Publish versions.yml
+
+When ready to release the new Spinnaker version to users, publish the
+`versions.yml` file to GCS bucket.
```
```suggestion
`versions.yml` file to GCS.
```

dbyron-sf commented on unittest/buildtool/versions_command_test.py:
```diff
@@ -0,0 +1,290 @@
+
+        got = builder.add_or_replace_release(release_111, releases)
+
+        self.assertEqual(len(got), 3)  # count of releases has not increased
```
This'll break if the max changes...maybe that's OK.  Maybe add an assertion above about `len(releases) == MAXIMUM_SPINNAKER_RELEASE_COUNT` and use `len(releases)` instead of 3?

kskewes-sf replied:
Should be okay. `add_or_replace_release()` doesn't enforce the length.
During `build()` we trim our "releases" list to `MAXIMUM_SPINNAKER_RELEASE_COUNT`

dbyron-sf commented on dev/buildtool/versions_commands.py:
```diff
@@ -0,0 +1,432 @@
+                          It is intended to support an "update" use case where
+                          only a subset of entries are updated within it.
+        """
+        self.__base_versions_dict = base_versions or {
```
I could use a comment about why this default value is helpful.  My reflex is to ditch the default and require `base_versions`, but I bet it's useful.

kskewes-sf replied:
I agree that it seems redundant to our general use case of updating the public GCS versions.yml file.
For extra context:

The `BomBuilder` and others all have a default dict which I have followed and rightly/wrongly it is easier than dealing with some none/nil/undefined behaviours.
I did miss the corresponding test case and have added in https://github.com/spinnaker/buildtool/pull/195/commits/87b42c2ca9f69c83f38fc33060d9c5bd39f1b2f8

That said, the `./buildtool.sh update_versions` command enforces passing in a `versions.yml` path which I thought was safer for anyone invoking buildtool. See: https://github.com/spinnaker/buildtool/pull/195/files#diff-9b53a58ad4a894dad2080a292e688a93b2f02c1fd93ed1c12043e6e0ee8fd63fR312.
So... there's a mismatch.

dbyron-sf replied:
Fair enough.  Let's ship it.

dbyron-sf commented on dev/buildtool/versions_commands.py:
```diff
@@ -0,0 +1,432 @@
+        minimum_halyard_version,
+        latest_halyard_version=None,
+    ):
+        """Build versions.yml. If an optional versions_dict is supplied then update."""
```
I don't see a way to supply `versions_dict` here.

kskewes-sf replied:
```suggestion
        """Build versions.yml."""
```

Sorry the description needed to be corrected after moving `versions_dict` to a constructor.

kskewes-sf replied:
And it is a very descriptive description... :joy:

kskewes-sf committed:
docs(ci): Improve instruction clarity

Co-authored-by: David Byron <82477955+dbyron-sf@users.noreply.github.com>

kskewes-sf committed:
feat(ci): Add versions.yml test for default case

Validate that the default versions_dict is used correctly.

Generally we are updating the GCS published versions.yml instead which
other tests exercise through use of the test fixture.

kskewes-sf committed:
docs(ci): Improve instruction clarity

Co-authored-by: David Byron <82477955+dbyron-sf@users.noreply.github.com>

kskewes-sf committed:
fix(ci): Correct method description to match code after refactor
